examples/ConfigurationStoreBackup/BackupAppConfigurationStore.cs

```diff
@@ -0,0 +1,216 @@
+        private static readonly string primaryStoreEndpoint = Environment.GetEnvironmentVariable("PrimaryStoreEndpoint");
+        private static readonly string secondaryStoreEndpoint = Environment.GetEnvironmentVariable("SecondaryStoreEndpoint");
+        private const string QueueName = "eventgridqueue";
+        private const int MaxMessagesToRead = 32;
+
+        [FunctionName("BackupAppConfigurationStore")]
+        public static void Run([TimerTrigger("0 */10 * * * *")]TimerInfo myTimer, ILogger log)
+        {
+            log.LogInformation($"C# Timer trigger function executed at: {DateTime.Now}");
+            if (string.IsNullOrEmpty(storageConnectionString))
+            {
+                log.LogError($"Please ensure that the environment variable 'AzureWebJobsStorage' is correctly set.");
+                return;
+            }
+            if (string.IsNullOrEmpty(primaryStoreEndpoint))
+            {
+                log.LogError($"Please ensure that the environment variable 'PrimaryStoreEndpoint' is set to your primary store endpoint.");
+                return;
+            }
+            if (string.IsNullOrEmpty(secondaryStoreEndpoint))
+            {
+                log.LogError($"Please ensure that the environment variable 'SecondaryStoreEndpoint' is set to your secondary store endpoint.");
+                return;
+            }
+
+            try
+            {
+                QueueClient queueClient = new QueueClient(storageConnectionString, QueueName);
+                ConfigurationClient primaryAppConfigClient = new ConfigurationClient(new Uri(primaryStoreEndpoint), new ManagedIdentityCredential());
+                ConfigurationClient secondaryAppConfigClient = new ConfigurationClient(new Uri(secondaryStoreEndpoint), new ManagedIdentityCredential());
+
+                // Block the thread so that another function instance cannot be triggered while this is still processing.
+                BackupAppConfigurationStoreAsync(queueClient, primaryAppConfigClient, secondaryAppConfigClient, log).Wait();
+            }
+            catch (UriFormatException ex)
+            {
+                log.LogError($"Please ensure that App Configuration store endpoints are valid URLs.\nException: {ex.Message}");
+            }
+            catch(AuthenticationFailedException ex)
+            {
+                log.LogError($"Failed to connect to App Configuration store with Managed Identity credentials.\nException: {ex.Message}");
+            }
+            catch (RequestFailedException ex) when (ex.ErrorCode == QueueErrorCode.QueueNotFound)
+            {
+                log.LogError($"Queue '{QueueName}' does not exist.\nException: {ex.Message}");
+            }
+            catch (RequestFailedException ex)
+            {
+                log.LogError($"Aborting App Configuration store backup.\nService request failed with the following exception: {ex}");
+            }
```

Comment:
None of these exceptions are transient errors. I'm leaning towards letting them bubble up and crash the app. Thoughts?

Reply:
I agree that we don't need to catch these exceptions. What about transient exceptions that cause the backup to fail? Should we also let these terminate the function abnormally, so the Azure Function shows the status of that instance as Failed in the Azure Portal?

Reply:
I am catching these exceptions in order to provide an actionable error message to the users in these cases. 
The only way to mark the function as "Failed" is to rethrow any exceptions that we catch after logging error message. In our code snippet, the only transient exceptions that we can ignore are coming from parsing queue message and deleting queue messages. 
Everything else coming from AppConfig is non-transient and we abort backup anyway. How about rethrowing those exceptions then?

Reply:
What will users get when the Function crashes? I suppose there are logs?

Reply:
Yes, there will be logs either way. If the function crashes, users have the option of acting on that failure using AppInsights or LogAnalytics.

Reply:
That sounds good enough for users to debug their code.

Reply:
All right. I realized I was not clear in my previous answer. I was only commenting on the exceptions here.

We need to have a common understanding of transient errors first. Errors we talk about here, for example, the URI does not have a correct format or storage queue cannot be found, are *not* transient because they will not fix themselves no matter how many times you try.

However, during the backup, you may occasionally fail to read queue messages, load config, or set config. If you try again, they may just work fine. These errors are considered transient because you can recover from them. We shouldn't let these errors crash the Function. The backup is aborted but the Function didn't fail. Most of the time, no attention needs to be taken either.

I hope this clarifies a little.

Reply:
Okay, I think I misunderstood earlier. I thought that any exceptions thrown by App Config are not transient because we are failing to backup config store and that's something users would like to be notified of. But actually the errors from App Config are transient because we may succeed in modifying config on a retry, and no action is needed from users.
The only exception I can think of where users need to take action is when a key being modified is locked in secondary store. In that case, backup will be stuck in an infinite loop unless someone unlocks the key from secondary store. Maybe we can catch that one exception and rethrow?